specs/eventing/data-plane.md

```diff
@@ -1,125 +1,172 @@
+content modes of the CloudEvents HTTP binding. Senders which do not advertise
+the ability to accept [reply events](#derived-reply-events) MAY implement only
+one content mode, as the recipient is not allowed to negotiate the content mode.
+Senders MUST support both cleartext (`http`) and TLS (`https`) URLs as event
+delivery destinations.
+
+### HTTP Verbs
+
+In the absence of specific delivery preferences, the sender MUST initiate
+delivery of the event to the recipient using the HTTP POST verb, using either
+the structured or binary encoding of the event (sender's choice). This delivery
+SHOULD be performed using the CloudEvents HTTP Binding, version 1.0.
+
+Senders MAY probe the recipient with an
+[HTTP OPTIONS request](https://tools.ietf.org/html/rfc7231#section-4.3.7); if
+implemented, the recipient MUST indicate support for the POST verb using the
+[`Allow` header](https://tools.ietf.org/html/rfc7231#section-7.4.1). Senders
+which receive an error when probing with HTTP OPTIONS SHOULD proceed using the
+HTTP POST mechanism.
+
+### Event Acknowledgement and Delivery Retry
+
+Event recipients MUST use the HTTP response code to indicate acceptance of an
+event. The recipient SHOULD NOT return a response accepting the event until it
+has handled event (processed the event or stored it in stable storage). The
+following response codes are explicitly defined; event recipients MAY also
+respond with other response codes. A response code not in this table SHOULD be
+treated as a retriable error.
+
+| Response code | Meaning                           | Retry | Delivery completed | Error |
+| ------------- | --------------------------------- | ----- | ------------------ | ----- |
+| `1xx`         | (Unspecified)                     | Yes\* | No\*               | Yes\* |
+| `200`         | [Event reply](#event-reply)       | No    | Yes                | No    |
+| `202`         | Event accepted                    | No    | Yes                | No    |
+| other `2xx`   | (Unspecified)                     | Yes\* | No\*               | Yes\* |
+| other `3xx`   | (Unspecified)                     | Yes\* | No\*               | Yes\* |
+| `400`         | Unparsable event                  | No    | No                 | Yes   |
+| `404`         | Endpoint does not exist           | Yes   | No                 | Yes   |
+| `409`         | Conflict / Processing in progress | Yes   | No                 | Yes   |
+| `429`         | Too Many Requests / Overloaded    | Yes   | No                 | Yes   |
+| other `4xx`   | Error                             | No    | No                 | Yes   |
+| other `5xx`   | Error                             | Yes   | No                 | Yes   |
+
+\* Unspecified `1xx`, `2xx`, and `3xx` response codes are **reserved for future
+extension**. Event recipients SHOULD NOT send these response codes in this spec
+version, but event senders MUST handle these response codes as errors and
+implement appropriate failure behavior.
+
```

Comment:
I just wanted to point out that retrying the unspecified 1xx, 2xx and all 3xx codes differs from what the distributed kafkachannel is doing today (in the `message_sender.go  SelectiveRetry()` logic @matzew previously linked).  It's not a big deal for us to switch to align with this but I wonder what the rationale is for retrying them?  Is it just that the "default" unexpected response code behavior errs on the side of "safety" by retrying?

- The 2xx codes are success codes so why does Knative need to be more restrictive and retry - would it the request be handled any differently?
- The 3xx (redirects) should be handled by the underlying http layer most likely right?  Would we expect a different result from retrying? (honest question - I'm not a networking guru ; )

As I said - just noting the difference with our current logic and we are ok to adapt if the spec changes.

Reply:
Yeah, I think those were just the "decisions" we made when implementing the KafkaChannel.  I was mostly just asking to spur the discussion in case other more opinionated folks wanted to weigh in ; ).  I've raised the idea in the past of making the set of response codes to retry on a configurable thing, but that's all theoretical at this point.  

The rewrite it looking great - thanks for all the effort!